File: questions/117. Populating Next Right Pointers in Each Node II/populating.py
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        cur = root
        prev = None
        head = None
        while cur:
            while cur:
                if cur.left:
                    if prev:
                        prev.next = cur.left
                    else:
                        head = cur.left
                    prev = cur.left
                if cur.right:
                    if prev:
                        prev.next = cur.right
                    else:
                        head = cur.right
                    prev = cur.right
                cur = cur.next
            cur = head
            prev = None
            head = None
        return root

File: questions/117. Populating Next Right Pointers in Each Node II/populating.java
class Solution {
    public Node connect(Node root) {
        Node head = null;
        Node cur = root;
        Node prev = null;
        while(cur != null){
            while(cur != null){
                if(cur.left != null){
                    if(prev != null){
                        prev.next = cur.left;
                    }else{
                        head = cur.left;
                    }
                    prev = cur.left;
                }
                if(cur.right != null){
                    if(prev != null){
                        prev.next = cur.right;
                    }else{
                        head = cur.right;
                    }
                    prev = cur.right;
                }
                cur = cur.next;
            }
            cur = head;
            prev = null;
            head = null;
        }
        return root;
    }
}
